Cython/Utility/StringTools.c

```diff
@@ -568,8 +506,13 @@ static CYTHON_INLINE PyObject* __Pyx_PyUnicode_Substring(
 static CYTHON_INLINE PyObject* __Pyx_PyUnicode_Substring(
             PyObject* text, Py_ssize_t start, Py_ssize_t stop) {
     Py_ssize_t length;
+#if CYTHON_ASSUME_SAFE_MACROS
     if (unlikely(__Pyx_PyUnicode_READY(text) == -1)) return NULL;
+#endif
```

Comment:
We must call this either way.
```suggestion
    if (unlikely(__Pyx_PyUnicode_READY(text) == -1)) return NULL;
```

Reply:
We can't - it isn't in the limited API, and the call is already in `PyUnicode_GetLength()` (https://github.com/python/cpython/blob/0f56adb8d74e703974811620559d96c999334547/Objects/unicodeobject.c#L4016) in versions old enough that it matters.